### ISSUE TITLE
rebuild necessary items based on changes to viz.yaml non-file options

### ISSUE THREAD

**Issue by aappling-usgs** [USGS-VIZLAB/vizlab#193]
first task: split the viz.yaml into pieces, write files behind the scenes that only include those pieces

second task: get dependencies hooked up so that items depend on their baby viz.yaml files

supports #103

**Comment by ldecicco-USGS** [comment]
OK, for step 1, I'm thinking of doing the save in `getContentInfo`...does that sound correct?

**Comment by aappling-usgs** [comment]
It needs to be called before the `make` targets that might depend on it; otherwise, they'll use old information. When & how would `getContentInfo` get called?

**Comment by ldecicco-USGS** [comment]
Ohhhhh right, I think that is post-make

**Comment by ldecicco-USGS** [comment]
Actually, if I run `createMakefiles` with `debug(getContentInfo)`, then I get pushed into the `getContentInfo` debugger, so I guess that is being called...

**Comment by aappling-usgs** [comment]
Just remember that `getContentInfo` also gets called lots of other times, e.g., every time you run a fetch/process/visualize function (via `as.viz`). 

What about creating a new target in the main Makefile that depends on viz.yaml and is depended on by fetch, process, etc.?

**Comment by ldecicco-USGS** [comment]
Talking myself through this....

I was thinking something like this:
```
getContentInfo <- function(viz.id, block=c('fetch','process','visualize','publish','resource'), no.match=c('stop','NA')) {
  if(length(viz.id) != 1) stop("viz.id must have length 1; see getContentInfos for alternatives")
  infos <- getContentInfos(viz.id, block=block)
  if(length(infos) == 0) {
    if(!missing(no.match) && is.na(no.match)) no.match <- 'NA'
    no.match <- match.arg(no.match)
    switch(
      no.match,
      'stop' = stop("no matches for viz.id '", paste0(viz.id, collapse="', '"), "' in block[s] ", paste0(block, collapse=", ")),
      'NA' = return(NA))
  }
  if(length(infos) > 1) stop("multiple matches for viz.id ", paste0(viz.id, collapse=", "), " in block[s] ", paste0(block, collapse=", "), ":\n", paste0("  ", names(infos), collapse="\n"))
  
  return.info <- checkConfig(info=infos[[1]], viz.id)
  
  return(infos[[1]])
}

checkConfig <- function(info,viz.id){
  file.config <- paste0(viz.id,".rds")
  full.config <- file.path("cache","config",file.config) 
  if(file.exists(full.config)){
    orig.info <- readRDS(full.config)
    if(!all(orig.info %in% info)){
      saveRDS(info, full.config)
    }
  } else {
    saveRDS(info, full.config)
  }
}
```
Then need to get the cache/config/`viz.id`.rds file pushed into the `make` files.

**Comment by aappling-usgs** [comment]
Ah, it's starting to become clearer! This is very helpful.

It looks like `checkConfig()` uses `getContentInfo()` for two things; (1) to get itself called, and (2) to use its output as the contents of the file to be checked/written. However, `getContentInfo()` gets called really often - more often than makefiles would ever use the information - so I think you can and should flip the order so that checkConfig gets itself called in another way and then makes a call to `getContentInfo()` to get the config block.

To serve purpose 1, making sure checkConfig() gets called, I'm increasingly liking the idea of a separate target in the top-level Makefile. That makefile gets used once per call to `make` - i.e., always at the right frequency for `make` to decide about updates. If it's a symbolic target (i.e., not a filename), and if the `fetch`, `process`, `visualize`, and `publish` targets all depend on it, then it will get run every time, which would be fine for a first cut. It could also write an indicator output file after checking all targets so that it only gets run as often as viz.yaml changes, which would be even more optimized.

**Comment by aappling-usgs** [comment]
I'd prefer another name for `checkConfig()` that communicates that the function often writes files, and that can't be read as "I confirm that the info is well formatted", which is how i first read it. What about `writeContentInfo` or `updateContentInfoFile` or something like that?

**Comment by ldecicco-USGS** [comment]
I really don't understand the idea of "a separate target in the top-level Makefile" (although that's exactly what I'm thinking for the global issue). Won't that cause the whole thing to rebuild if any part of the yaml is changed?

**Comment by aappling-usgs** [comment]
It will cause the whole thing to reconsider building, yes, which I think is what we want. But it won't necessarily cause every step to be refetched or reprocessed or anything. By "reconsider" I mean it'll check timestamps for each of those fetch/process/visualize/publish targets - if the content info file for any particular target has a more recent timestamp than the target, then that target and its downstream dependencies will be rebuilt. but if you didn't change the timestamp on the content info file, then the target shouldn't get touched.

**Comment by aappling-usgs** [comment]
https://github.com/USGS-VIZLAB/vizlab/pull/206

**Comment by aappling-usgs** [comment]
Further needs for this issue are coming up as I'm playing with USGS-VIZLAB/example.

1. Can we make it so `config` is a dependency of more targets than just `all`?  For example, I'd like to be able to edit the viz.yaml, run `make fetch`, and see my updates take effect. (Really really ideally, I'd be able to edit the viz.yaml, run `make -f vizlab/make/fetch.make iris-data`, and see my updates...but that would require every single target to depend on config, which seems like a headache...but maybe worth it?) Whatever we choose, I think we only need the dependency at the finest granularity where we have the dependency - i.e., if we have `vizlab/make/fetch.make` depend on `config`, then `all` doesn't need to depend on `config`.

1. Can we make `config` be a file that gets written (probably tucked into vizlab/make somewhere) so that it only gets updated when `viz.yaml` has changed? This will help prevent the above change[s] from slowing down the build process a bunch.

1. Optional change, but now that I'm thinking about making the dependency more granular, I'm wondering which is faster from a vizlab user perspective: to only build the config .rds file for the specific target[s] being requested, just in time to build them, or to build all config .rds files the first time any one of them is noticed to be out of date (as currently implemented)? I guess this comes down to whether the maximum build time for a complex project is a noticeable wait or not - it might always be too fast to matter.

1. There's a scenario where having all fetch items depend on their respective config files is a problem: when the fetch item is a file in `data` and therefore never gets copied into the cache, its timestamp never changes. So if it depends on cache/config/xx.rds, it gets 'fetched' every time. The fetch is just a few shell & function calls, no real action, but it still adds to every build time by a second or two. My example is iris-data in USGS-VIZLAB/example; it may be less common in real vizzies, but it didn't take long to find this real example: https://github.com/USGS-VIZLAB/water-use/blob/master/viz.yaml#L121.

1. I'm noticing that we save internal timestamp-checking files to vizlab/make/timestamps. Would it make sense to move `cache/config/xx.rds` over into `vizlab/make/config/xx.rds`? i don't feel strongly about this, but it might be nice to hide those .rds files from view since users shouldn't edit or commit them.

**Comment by ldecicco-USGS** [comment]
1. on it. I thought if all depended on it, then it would cascade, but I understand now.
2. I don't follow what you are asking. I can't fathom that we're adding much time to writing these tiny files. 
3. Again....I really can't fathom that building these new files and then checking if they are different is taking enough time that anyone would notice relative to the rest of the viz...but haven't benchmarked. If it somehow adds a noticeable affect, then we could look into using `feather`, but, (a) they are generally only made once, and (b) they are tiny tiny files....So, 
4. Yeah, I don't understand the fetch timestamping stuff at all. I'll try re-reading this after I do (1)
5. I think it would make sense to move those. That is, hiding all the "vizlab-logistics" files in a single place.

**Comment by ldecicco-USGS** [comment]
So for (1) above, I think you are saying go back to the original Makefile:
```
all: \
	fetch \
	process \
	visualize \
	publish

```
Then the simpler way:

add to each fetch.make, process.make, etc:
```
all: \
	config \	
	iris-data \
	cars-data \
	Cuyahoga \
	MakeMessages

config: viz.yaml
	export R_LIBS_USER=$(RLIBSUSER);\
	${RSCRIPT}  -e 'library(vizlab); updateConfig()'

iris-data: data/iris2.csv
data/iris2.csv: vizlab/make/timestamps/iris-data cache/config/iris-data.rds
...
```

Or the mo'better way, something like this??:
```
all: \
	config_iris-data \	
	iris-data \
	config_cars-data \
	cars-data 

config_iris-data: viz.yaml
	export R_LIBS_USER=$(RLIBSUSER);\
	${RSCRIPT}  -e 'library(vizlab); updateConfig(iris-data)'
iris-data: data/iris2.csv
data/iris2.csv: vizlab/make/timestamps/iris-data cache/config/iris-data.rds
	export R_LIBS_USER=$(RLIBSUSER);\
	${RBATCH} "--args fun=fetch funargs=list(viz='iris-data')" \
	vizlab/make/callFunction.R vizlab/make/log/fetch/iris-data.Rout

config_cars-data: viz.yaml
	export R_LIBS_USER=$(RLIBSUSER);\
	${RSCRIPT}  -e 'library(vizlab); updateConfig(cars-data)'
cars-data: cache/cars.csv
cache/cars.csv: vizlab/make/timestamps/cars-data cache/config/cars-data.rds scripts/fetch/cars.R
	export R_LIBS_USER=$(RLIBSUSER);\
	${RBATCH} "--args fun=fetch funargs=list(viz='cars-data') scripts=c('scripts/fetch/cars.R')" \
	vizlab/make/callFunction.R vizlab/make/log/fetch/cars-data.Rout
vizlab/make/timestamps/cars-data: cars-data_timestamp
cars-data_timestamp: scripts/fetch/cars.R
	export R_LIBS_USER=$(RLIBSUSER);\
	${RBATCH} "--args fun=fetchTimestamp funargs=list(viz='cars-data') scripts=c('scripts/fetch/cars.R')" \
	vizlab/make/callFunction.R vizlab/make/log/fetch/cars-data_timestamp.Rout
```

**Comment by aappling-usgs** [comment]
1,5 - sounds good.
2. In the case where updateConfig() checks all files every time and is depended on by multiple targets, then a single call to `make all` would call updateConfig maybe 4 times, maybe 30 times (depending on where you put the dependencies) - having updateConfig write a single tiny file once it's complete would cut that number down to exactly once per `make` call, and done.
3. I think you're actually headed this way with your mo'better approach above: iris_data depends on its own personal iris-data.rds and updateConfig(iris-data), rather than having just one or a few calls to updateConfig at the top levels (all, fetch, etc.) that update all the configs all at once. We might need to make some implementation tweaks to the above, but it's close.
4. I think the cleanest fix might be on the target side, separate from this config-writing task, even though it emerged here. Let's just get it on the table and talk about it once this issue (and maybe also globals) is done.

**Comment by aappling-usgs** [comment]
Tweaks: just one, actually: I think we could do with a single generic target `cache/config/%.rds` that would define dependencies of each individual config/xx.rds on viz.yaml and updateConfig(). For now we'd need to copy it so it appears once in fetch.make, once in process.make, etc., but with a little further adjusting (for https://github.com/USGS-VIZLAB/vizlab/issues/88) we can probably collapse it down to just a single appearance.

fetch.make
```make
cache/config/%.rds : viz.yaml
	${RSCRIPT}  -e 'library(vizlab); updateConfig("$*")'

iris-data: data/iris.csv
data/iris.csv: cache/config/iris-data.rds
	export R_LIBS_USER=$(RLIBSUSER);\
	${RBATCH} "--args fun=fetch funargs=list(viz='iris-data')" \
	vizlab/make/callFunction.R vizlab/make/log/fetch/iris-data.Rout

cars-data: cache/cars.csv
cache/cars.csv: vizlab/make/timestamps/cars-data cache/config/cars-data.rds scripts/fetch/cars.R
	export R_LIBS_USER=$(RLIBSUSER);\
	${RBATCH} "--args fun=fetch funargs=list(viz='cars-data') scripts=c('scripts/fetch/cars.R')" \
	vizlab/make/callFunction.R vizlab/make/log/fetch/cars-data.Rout
vizlab/make/timestamps/cars-data: cars-data_timestamp
cars-data_timestamp: scripts/fetch/cars.R
	export R_LIBS_USER=$(RLIBSUSER);\
	${RBATCH} "--args fun=fetchTimestamp funargs=list(viz='cars-data') scripts=c('scripts/fetch/cars.R')" \
	vizlab/make/callFunction.R vizlab/make/log/fetch/cars-data_timestamp.Rout
```

$* is an "automatic variable" in make - see https://www.gnu.org/software/make/manual/html_node/Automatic-Variables.html. It gives the stem of the target, which in this case should be things like iris-data and cars-data

**Comment by aappling-usgs** [comment]
made a new issue (#207, linked just now) to capture number 4 above.

**Comment by aappling-usgs** [comment]
The core issue is addressed thoroughly. Ideas 1, 3, and 5 above are addressed thoroughly. Point 4 now has its own issue and point 2 is partly addressed by #208, partly tabled until build time for this implementation becomes concerning (which may not ever happen). So I think we're good to go!
